### PR TITLE
[CPU] Introduce vector unit dims optimization in vector lowering.

### PR DESCRIPTION
**Pull request by hanhanW**
The revision is mainly for vector transfer optimization. It also applies transfer ops write to read forwarding and dead transfer write optimizations before optimizing vector unit dims. Removing unit dims (especially for trailing unit dims) is very important to CPU targets. The progressive lowering will unroll vector.transfer ops to 1D vector.load/store. If the trailing unit dims are not folded, it will generate bunch of scalar loads and stop further optimization (e.g., transpose lowering).

---------

Co-authored-by: Jerry Wu <cheyuw@google.com>

benchmark-extra: x86_64-dt-only

### CODE REVIEW

**Comment by hanhanW** [comment]
Okay, there are issues about masking if we enable casting leading unit dims patterns..

**Comment by pzread** [comment]
> Okay, there are issues about masking if we enable casting leading unit dims patterns..

I believe this change is related: https://github.com/llvm/llvm-project/pull/76468

**Comment by hanhanW** [comment]
> > Okay, there are issues about masking if we enable casting leading unit dims patterns..
> 
> I believe this change is related: [llvm/llvm-project#76468](https://github.com/llvm/llvm-project/pull/76468)

oh, thanks! I will try it

**Comment by hanhanW** [comment]
@pzread do you know why I can't compare the benchmark results with parent commit? I think I rebased on https://github.com/openxla/iree/commit/e82cfa99b3dc6447fe926329c12414bce75c0507 and the result of parent commit should be already pushed to the dashboard? https://github.com/openxla/iree/actions/runs/7610466940/job/20735638558

**Comment by pzread** [comment]
> @pzread do you know why I can't compare the benchmark results with parent commit? I think I rebased on [e82cfa9](https://github.com/openxla/iree/commit/e82cfa99b3dc6447fe926329c12414bce75c0507) and the result of parent commit should be already pushed to the dashboard? https://github.com/openxla/iree/actions/runs/7610466940/job/20735638558

I found the problem... The workflow https://github.com/openxla/iree/actions/runs/7610466940/attempts/2 only uploads the partial results due to a tricky bug with partial re-run.

You can rebase this PR again and it should retrieve the complete results from a later commit

**Comment by hanhanW** [comment]
I see.. this is tricky. Thanks for looking at it!

**Comment by hanhanW** [comment]
The performance regression about GPT2_117M_1X4 on pixel looks like a noise. The total dispatch size does no change.

image

**Comment by hanhanW** [comment]
The total dispatch size regressions are mostly about packing with unit inner tile sizes, I filed an issue for tracking it. It should be an easy fix: https://github.com/openxla/iree/issues/16181

**Comment by hanhanW** [comment]
This requires an upstream commit (authored by @pzread ): https://github.com/llvm/llvm-project/commit/dedc7d4d362b8045c6810f8ca7f947bbdb63b7ec  I'll wait for it, but I want to open the PR for gathering some feedback.

**Comment by hanhanW** [comment]
I verified that the size regressions are fixed (and even improved) in https://github.com/openxla/iree/pull/16183 The regression issues are addressed by pack/unpack simplification. I think this is the approach we want generally for CPU backends.

**Review by bjacob — APPROVED** [review]

**Review by pzread — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUOptimizeVectorUnitDims.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMCPU/PassDetail.h"
+#include "iree/compiler/Codegen/LLVMCPU/Passes.h"
+#include "mlir/Dialect/Vector/Transforms/LoweringPatterns.h"
+#include "mlir/Dialect/Vector/Transforms/VectorTransforms.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-llvmcpu-optimize-vector-unit-dims"
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUOptimizeVectorUnitDims.cpp]: %s/optimize/drop/g? :)

**Review by dcaballe — APPROVED** [review]
Thanks for adding this pass! I think it will be very useful as we may have to invoke it multiple times. Just a minor naming comment but LGTM.

**Pull request merged**